### DISCUSSION TITLE
disabling DeePMD TF32 calculation in LAMMPS

### DISCUSSION THREAD

**Discussion by yjw0510** [deepmodeling/deepmd-kit#1920] [Q&A] [answered]
Dear DeePMD users,

I'm trying to use my DeePMD model to do molecular dynamics simulation in LAMMPS.

I have two different types of GPUs (RTX 3090 and A100).

I found that both GPUs support TF32 calculations.
However, the lack of precision does make an undesired effect in the simulation so I want to disable them.

Generally, `NVIDIA_TF32_OVERRIDE=0` should turn off TF32 but it seems Tensorflow does not do that.
(https://servicedesk.surf.nl/wiki/display/WIKI/Deep+Learning+on+A100+GPUs)
<img width="1135" alt="image" src="https://user-images.githubusercontent.com/26711298/190862325-1d60981f-4521-4055-a16d-79ace9de9029.png">

I'm using a compressed model using `DP_INTERFACE_PREC=low` 

**Slurm message**
<img width="1512" alt="image" src="https://user-images.githubusercontent.com/26711298/190863053-fea3671c-f939-4b06-85a8-238db8431bb2.png">

with `CUBLASLT_LOG_LEVEL=5`, I found the log below.
<img width="1512" alt="image" src="https://user-images.githubusercontent.com/26711298/190863087-0d9e8d15-edf1-4122-adfa-fc4a1d871f3f.png">

**Comment by njzjz** [comment, marked as answer]
For C++ applications, one needs to call `tensorflow::enable_tensor_float_32_execution(false)` to disable tensor float 32.
See https://github.com/tensorflow/tensorflow/blob/5dcfc51118817f27fad5246812d83e5dccdc5f72/tensorflow/core/platform/tensor_float_32_utils.h#L21.

**Reply by yjw0510** [reply]
Thank you for the answer. Then I think the only way is to recompile the code with the `tensorflow::enable_tensor_float_32_execution(false)` Is there any way to disable it without recompiling and without touching the original DeePMD code as much as possible?

**Reply by njzjz** [reply]
LAMMPS provide a way to load a external DSO, where you can call TF. See https://docs.lammps.org/plugin.html
